Create state description merging service

lpapazow opened the pull request:
Created as a fix for #2267.

As suggested in this [discussion](https://github.com/eclipse/smarthome/pull/4428#issuecomment-340480706), created a **StateDescriptionService** that has a **getStateDescription** method, which iterates over all **StateDescriptionProviders** and sets the **StateDescription**'s fields if they haven't been set by a **StateDescriptionProvider** with higher ranking.

The list of **stateDescriptionProviders** is moved from the **ItemRegistryImpl** class to the **StateDescriptionService** and a **StateDescriptionService** reference is added to the **ItemRegistryImpl** which is later passed to the **GenericItem** class. 

Changed **GenericItem**'s **getStateDescription** method implementation to use **StateDescriptionService**'s **getStateDescription** method.

Added tests for the StateDescription service in the **StateDescriptionServiceImplTest** class.


Signed-off-by: Lyubomir V. Papazov <lpapazow@gmail.com>

triller-telekom commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+
+    private static boolean shouldCheckNextDescriptionProvider(BigDecimal minimum, BigDecimal maximum, BigDecimal step,
+            String pattern, boolean isReadOnly, List<StateOption> stateOptions) {
+        return minimum == null || maximum == null || step == null || pattern == null || isReadOnly == false
```
Should we document somewhere that if you want to create a "winning" `StateDescriptionProvider` that you have to set not only its priority to the highest value but **in addition* also provide values for **all* fields, i.e. minimum, maximum, step, pattern, isReadOnly + stateOptions?

htreu replied:
Imho it there should not be such a "winning" ConfigDescriptionProvider: The Channel config description is always to be taken first, so if this one defines all fields from the beginning other would never be able to overwrite. This would result in the pattern from *.items files or *.sitemap files never to be taken into account.

My impression is that this PR should at first just introduce the new StateDescriptionService without changing the merging logic.

triller-telekom commented on bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImplTest.java:
```diff
@@ -0,0 +1,265 @@
+		StateDescription finalStateDescription = item.getStateDescription();
+
+		assertThat(
+				"expected minimum to be equal to the minumum from provider one because it has higher service ranking",
```
Can you please remove the explanation string in the `assertThat` calls because they are hard to keep in sync and most of the time they are either stating the obvious or if they are meant as a error message for the failing test they are written to state the opposite of what the test actually checks.

triller-telekom commented on bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImplTest.java:
```diff
@@ -0,0 +1,265 @@
+		String stateDescriptionProvider1Pattern = "pattern1";
+		String stateDescriptionProvider2Pattern = "pattern2";
+
+		StateDescriptionProvider stateDescriptionProvider1 = mock(StateDescriptionProvider.class);
```
Since the following lines down to `StateDescription finalStateDescription = item.getStateDescription();` are always the same for each test, how about creating a function that you pass the rank and description form 2 providers and return the `finalStateDescription`, to avoid the duplication of the code?

htreu commented on bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImplTest.java:
```diff
@@ -0,0 +1,265 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
```
empty line

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/items/ItemRegistryImpl.java:
```diff
@@ -388,31 +388,14 @@ protected void activate(final ComponentContext componentContext) {
-        }
+    
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.DYNAMIC)
+    protected void addStateDescriptionMergingService(StateDescriptionService stateDescriptionService) {
```
since we now have only one service which is injected & the service name is `StateDescriptionService` please rename the method to `setStateDescriptionService`.

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/items/ItemRegistryImpl.java:
```diff
@@ -388,31 +388,14 @@ protected void activate(final ComponentContext componentContext) {
-            }
-        }
+    
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.DYNAMIC)
```
The new service should have `cardinality=OPTIONAL` and then should be given to all existing items when set (just like a new StateDescriptionProvider before).

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/items/ItemRegistryImpl.java:
```diff
@@ -388,31 +388,14 @@ protected void activate(final ComponentContext componentContext) {
-            ((GenericItem) item).setStateDescriptionProviders(stateDescriptionProviders);
-        }
+    
+    protected void removeStateDescriptionMergingService(StateDescriptionService stateDescriptionService) {
```
please rename the method to `unsetStateDescriptionService` and also unset the StateDescriptionService` from all available items.

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionService.java:
```diff
@@ -0,0 +1,43 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
```
remove empty line

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
```
remove empty line

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
```
Please give a short explanation what this component is doing.

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+            StateDescription currentDescription = stateDescriptionProvider.getStateDescription(itemName, locale);
+            if (currentDescription != null) {
+                foundNonNullStateDescription = true;
+                if (shouldCheckNextDescriptionProvider(minimum, maximum, step, pattern, isReadOnly, stateOptions)) {
```
matter of taste, but i would like this to be turned around, like `if(allPropertiesSet(...)) break;` and otherwise continue with the overwrite logic.

htreu replied:
Besides that: the new merge logic might result in different behaviour and makes this API breaking I guess. wdyt?

htreu commented:
Hey @lpapazow any update on this?

lpapazow commented:
I updated the PR, sorry for the delay. 

As suggested [in your comment](https://github.com/eclipse/smarthome/pull/4882#discussion_r161442979), I reverted the changes regarding the merging logic in `getStateDescription` and the PR now just introduces the new `StateDescriptionService`. 

However, the current version of this PR does not resolve the [orinial issue](https://github.com/eclipse/smarthome/issues/2267) that I was trying to fix.

@htreu Can you please start a discussion about an alternative way to merge StateDecriptions without losing relevant information from lower ranking description providers once this PR is merged.

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/items/ItemRegistryImpl.java:
```diff
@@ -159,9 +156,9 @@ private void replaceInGroupItems(Item oldItem, Item newItem, List<String> groupI
      * An item should be initialized, which means that the event publisher is
      * injected and its implementation is notified that it has just been
-     * created, so it can perform any task it needs to do after its creation.
+     *  created, so it can perform any task it needs to do after its creation.
```
this change (adding a blank character) should be reverted.

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/items/ItemRegistryImpl.java:
```diff
@@ -159,9 +156,9 @@ private void replaceInGroupItems(Item oldItem, Item newItem, List<String> groupI
+     *  created, so it can perform any task it needs to do after its creation.
      *
-     * @param item the item to initialize
+     * @param item  the item to initialize
```
this change (adding a blank character) should be reverted.

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GenericItem.java:
```diff
@@ -164,9 +163,9 @@ public void removeGroupName(String groupItemName) {
-
-    public void setStateDescriptionProviders(List<StateDescriptionProvider> stateDescriptionProviders) {
-        this.stateDescriptionProviders = stateDescriptionProviders;
+    
```
please use the default formatter or remove the blank chars manually.

htreu commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionService.java:
```diff
@@ -0,0 +1,43 @@
+     * multiple StateDescriptionProviders into one final StateDescription
+     * 
+     * @param itemName
+     * 
```
Please sue the default ESH formatter here once to get rid of the obsolete empty lines.

lpapazow commented:
While rebasing the master branch, I noticed that #5022 introduces `Nullable` annotations for the `stateDescriptionProviders` and `StateDescription` in the `GenericItem` class. 
Should I add such annotations for the `stateDescriptionService` filed in the `GenericItem` class?

htreu commented:
@lpapazow yes, please add null annotations, thanks.

Regarding the original issue #2267: With your changes here any solution is now able to provide its own implementation of a StateDescriptionService. OSGi DS should then replace the ESH provided service with the custom solution service if its service rank is higher. For pure ESH this means the StateDescription calculation will not change.

RFC: @kaikreuzer, @sjka, @maggu2810

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionService.java:
```diff
@@ -0,0 +1,42 @@
+     * 
+     * @param itemName
+     * @param locale (can be null)
+     * 
```
Remove that empty line

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionService.java:
```diff
@@ -0,0 +1,42 @@
+     * Implementations of this method merge the StateDescriptions provided from
+     * multiple StateDescriptionProviders into one final StateDescription
+     * 
+     * @param itemName
```
Add description (also if it is pretty clear) to the parameters.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionService.java:
```diff
@@ -0,0 +1,42 @@
+public interface StateDescriptionService {
+
+    /**
+     * Implementations of this method merge the StateDescriptions provided from
```
Shouldn't the first comment end with a dot?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GenericItem.java:
```diff
@@ -370,32 +369,7 @@ public void setCategory(@Nullable String category) {
-        }
-
-        return result;
+        return stateDescriptionService.getStateDescription(this.name, locale);
```
This could result in an NPE as stateDescriptionService is nullable (and a dynamic+optional dependency in ItemRegistryImpl).

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.service;
```
Good point
